File: go/volume.go
package modal

import (
	"context"
	"fmt"

	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
)

// VolumeService provides Volume related operations.
type VolumeService interface {
	FromName(ctx context.Context, name string, params *VolumeFromNameParams) (*Volume, error)
	Ephemeral(ctx context.Context, params *VolumeEphemeralParams) (*Volume, error)
	Delete(ctx context.Context, name string, params *VolumeDeleteParams) error
}

type volumeServiceImpl struct{ client *Client }

// Volume represents a Modal Volume that provides persistent storage.
type Volume struct {
	VolumeID        string
	Name            string
	readOnly        bool
	cancelEphemeral context.CancelFunc
}

// VolumeFromNameParams are options for finding Modal Volumes.
type VolumeFromNameParams struct {
	Environment     string
	CreateIfMissing bool
}

// FromName references a Volume by its name.
func (s *volumeServiceImpl) FromName(ctx context.Context, name string, params *VolumeFromNameParams) (*Volume, error) {
	if params == nil {
		params = &VolumeFromNameParams{}
	}

	creationType := pb.ObjectCreationType_OBJECT_CREATION_TYPE_UNSPECIFIED
	if params.CreateIfMissing {
		creationType = pb.ObjectCreationType_OBJECT_CREATION_TYPE_CREATE_IF_MISSING
	}

	resp, err := s.client.cpClient.VolumeGetOrCreate(ctx, pb.VolumeGetOrCreateRequest_builder{
		DeploymentName:     name,
		EnvironmentName:    environmentName(params.Environment, s.client.profile),
		ObjectCreationType: creationType,
	}.Build())

	if status, ok := status.FromError(err); ok && status.Code() == codes.NotFound {
		return nil, NotFoundError{fmt.Sprintf("Volume '%s' not found", name)}
	}
	if err != nil {
		return nil, err
	}

	s.client.logger.DebugContext(ctx, "Retrieved Volume", "volume_id", resp.GetVolumeId(), "volume_name", name)
	return &Volume{VolumeID: resp.GetVolumeId(), Name: name, readOnly: false, cancelEphemeral: nil}, nil
}

// ReadOnly configures Volume to mount as read-only.
func (v *Volume) ReadOnly() *Volume {
	return &Volume{
		VolumeID:        v.VolumeID,
		Name:            v.Name,
		readOnly:        true,
		cancelEphemeral: v.cancelEphemeral,
	}
}

// IsReadOnly returns true if the Volume is configured to mount as read-only.
func (v *Volume) IsReadOnly() bool {
	return v.readOnly
}

// VolumeEphemeralParams are options for client.Volumes.Ephemeral.
type VolumeEphemeralParams struct {
	Environment string
}

// VolumeDeleteParams are options for client.Volumes.Delete.
type VolumeDeleteParams struct {
	Environment  string
	AllowMissing bool
}

// Ephemeral creates a nameless, temporary Volume, that persists until CloseEphemeral is called, or the process exits.
func (s *volumeServiceImpl) Ephemeral(ctx context.Context, params *VolumeEphemeralParams) (*Volume, error) {
	if params == nil {
		params = &VolumeEphemeralParams{}
	}

	resp, err := s.client.cpClient.VolumeGetOrCreate(ctx, pb.VolumeGetOrCreateRequest_builder{
		ObjectCreationType: pb.ObjectCreationType_OBJECT_CREATION_TYPE_EPHEMERAL,
		EnvironmentName:    environmentName(params.Environment, s.client.profile),
	}.Build())
	if err != nil {
		return nil, err
	}

	s.client.logger.DebugContext(ctx, "Created ephemeral Volume", "volume_id", resp.GetVolumeId())

	ephemeralCtx, cancel := context.WithCancel(context.Background())
	startEphemeralHeartbeat(ephemeralCtx, func() error {
		_, err := s.client.cpClient.VolumeHeartbeat(ephemeralCtx, pb.VolumeHeartbeatRequest_builder{
			VolumeId: resp.GetVolumeId(),
		}.Build())
		return err
	})

	return &Volume{
		VolumeID:        resp.GetVolumeId(),
		readOnly:        false,
		cancelEphemeral: cancel,
	}, nil
}

// CloseEphemeral deletes an ephemeral Volume, only used with VolumeEphemeral.
func (v *Volume) CloseEphemeral() {
	if v.cancelEphemeral != nil {
		v.cancelEphemeral()
	} else {
		// We panic in this case because of invalid usage. In general, methods
		// used with `defer` like CloseEphemeral should not return errors.
		panic(fmt.Sprintf("Volume %s is not ephemeral", v.VolumeID))
	}
}

// Delete deletes a named Volume.
//
// Warning: Deletion is irreversible and will affect any Apps currently using the Volume.
func (s *volumeServiceImpl) Delete(ctx context.Context, name string, params *VolumeDeleteParams) error {
	if params == nil {
		params = &VolumeDeleteParams{}
	}

	volume, err := s.FromName(ctx, name, &VolumeFromNameParams{
		Environment:     params.Environment,
		CreateIfMissing: false,
	})

	if err != nil {
		if _, ok := err.(NotFoundError); ok && params.AllowMissing {
			return nil
		}
		return err
	}

	_, err = s.client.cpClient.VolumeDelete(ctx, pb.VolumeDeleteRequest_builder{
		VolumeId: volume.VolumeID,
	}.Build())

	if err != nil {
		if st, ok := status.FromError(err); ok && st.Code() == codes.NotFound && params.AllowMissing {
			return nil
		}
		return err
	}

	s.client.logger.DebugContext(ctx, "Deleted Volume", "volume_name", name, "volume_id", volume.VolumeID)
	return nil
}

File: js/src/volume.ts
import { ObjectCreationType } from "../proto/modal_proto/api";
import { getDefaultClient, type ModalClient } from "./client";
import { ClientError, Status } from "nice-grpc";
import { NotFoundError, InvalidError } from "./errors";
import { EphemeralHeartbeatManager } from "./ephemeral";

/** Optional parameters for {@link VolumeService#fromName client.volumes.fromName()}. */
export type VolumeFromNameParams = {
  environment?: string;
  createIfMissing?: boolean;
};

/** Optional parameters for {@link VolumeService#ephemeral client.volumes.ephemeral()}. */
export type VolumeEphemeralParams = {
  environment?: string;
};

/** Optional parameters for {@link VolumeService#delete client.volumes.delete()}. */
export type VolumeDeleteParams = {
  environment?: string;
  allowMissing?: boolean;
};

/**
 * Service for managing {@link Volume}s.
 *
 * Normally only ever accessed via the client as:
 * ```typescript
 * const modal = new ModalClient();
 * const volume = await modal.volumes.fromName("my-volume");
 * ```
 */
export class VolumeService {
  readonly #client: ModalClient;
  constructor(client: ModalClient) {
    this.#client = client;
  }

  /**
   * Reference a {@link Volume} by its name.
   */
  async fromName(name: string, params?: VolumeFromNameParams): Promise<Volume> {
    try {
      const resp = await this.#client.cpClient.volumeGetOrCreate({
        deploymentName: name,
        environmentName: this.#client.environmentName(params?.environment),
        objectCreationType: params?.createIfMissing
          ? ObjectCreationType.OBJECT_CREATION_TYPE_CREATE_IF_MISSING
          : ObjectCreationType.OBJECT_CREATION_TYPE_UNSPECIFIED,
      });
      this.#client.logger.debug(
        "Retrieved Volume",
        "volume_id",
        resp.volumeId,
        "volume_name",
        name,
      );
      return new Volume(resp.volumeId, name);
    } catch (err) {
      if (err instanceof ClientError && err.code === Status.NOT_FOUND)
        throw new NotFoundError(err.details);
      throw err;
    }
  }

  /**
   * Create a nameless, temporary {@link Volume}.
   * It persists until closeEphemeral() is called, or the process exits.
   */
  async ephemeral(params: VolumeEphemeralParams = {}): Promise<Volume> {
    const resp = await this.#client.cpClient.volumeGetOrCreate({
      objectCreationType: ObjectCreationType.OBJECT_CREATION_TYPE_EPHEMERAL,
      environmentName: this.#client.environmentName(params.environment),
    });

    this.#client.logger.debug(
      "Created ephemeral Volume",
      "volume_id",
      resp.volumeId,
    );

    const ephemeralHbManager = new EphemeralHeartbeatManager(() =>
      this.#client.cpClient.volumeHeartbeat({ volumeId: resp.volumeId }),
    );

    return new Volume(resp.volumeId, undefined, false, ephemeralHbManager);
  }

  /**
   * Delete a named {@link Volume}.
   *
   * Warning: Deletion is irreversible and will affect any Apps currently using the Volume.
   */
  async delete(name: string, params?: VolumeDeleteParams): Promise<void> {
    try {
      const volume = await this.fromName(name, {
        environment: params?.environment,
        createIfMissing: false,
      });
      await this.#client.cpClient.volumeDelete({
        volumeId: volume.volumeId,
      });
      this.#client.logger.debug(
        "Deleted Volume",
        "volume_name",
        name,
        "volume_id",
        volume.volumeId,
      );
    } catch (err) {
      const isNotFound =
        err instanceof NotFoundError ||
        (err instanceof ClientError && err.code === Status.NOT_FOUND);
      if (isNotFound && params?.allowMissing) {
        return;
      }
      throw err;
    }
  }
}

/** Volumes provide persistent storage that can be mounted in Modal {@link Function_ Function}s. */
export class Volume {
  readonly volumeId: string;
  readonly name?: string;
  private _readOnly: boolean = false;
  readonly #ephemeralHbManager?: EphemeralHeartbeatManager;

  /** @ignore */
  constructor(
    volumeId: string,
    name?: string,
    readOnly: boolean = false,
    ephemeralHbManager?: EphemeralHeartbeatManager,
  ) {
    this.volumeId = volumeId;
    this.name = name;
    this._readOnly = readOnly;
    this.#ephemeralHbManager = ephemeralHbManager;
  }

  /**
   * @deprecated Use {@link VolumeService#fromName client.volumes.fromName()} instead.
   */
  static async fromName(
    name: string,
    options?: VolumeFromNameParams,
  ): Promise<Volume> {
    return getDefaultClient().volumes.fromName(name, options);
  }

  /** Configure Volume to mount as read-only. */
  readOnly(): Volume {
    return new Volume(this.volumeId, this.name, true, this.#ephemeralHbManager);
  }

  get isReadOnly(): boolean {
    return this._readOnly;
  }

  /**
   * @deprecated Use {@link VolumeService#ephemeral client.volumes.ephemeral()} instead.
   */
  static async ephemeral(options: VolumeEphemeralParams = {}): Promise<Volume> {
    return getDefaultClient().volumes.ephemeral(options);
  }

  /** Delete the ephemeral Volume. Only usable with emphemeral Volumes. */
  closeEphemeral(): void {
    if (this.#ephemeralHbManager) {
      this.#ephemeralHbManager.stop();
    } else {
      throw new InvalidError("Volume is not ephemeral.");
    }
  }
}
